### PR TITLE
fix assertion bug for SWA API in TE-JAX

### PR DESCRIPTION
**Pull request by kocchop**
# Description

There is an assertion bug in TE-JAX. This PR addresses the issue

Fixes # bug in assertion logic for SWA in TE-JAX

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- fixed the `assert` logic in `FusedAttnCPWithAllGatherFwdPrimitive` and `FusedAttnCPWithAllGatherBwdPrimitive`

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/jax/cpp_extensions/attention.py
Comment (ksivaman) [inline transformer_engine/jax/cpp_extensions/attention.py:1060]: I'm not sure I understand this change. This assertion is meant to catch the case when sliding window attention is being used with Context parallelism, so the existing check is correct. Could you post some more details (and a stacktrace if possible) about your bug?

**Review by ksivaman — CHANGES_REQUESTED** [review]
Attached comment

File: transformer_engine/jax/cpp_extensions/attention.py
Comment (huanghua1994) [reply transformer_engine/jax/cpp_extensions/attention.py:1060]: Please have a look at this: https://github.com/NVIDIA/TransformerEngine/pull/1205#discussion_r1776149416

Comment (kocchop) [reply transformer_engine/jax/cpp_extensions/attention.py:1060]: Hi @ksivaman , the current assert logic is becoming false for all non CP related runs, hence blocking them. I am running without CP and without SWA but still seeing the `AssertionError`:
```
...
number parameters: 6.738 billion
Per train step:
 Total TFLOPs: 755.06 
 split as 86.02% learnable weight flops and 13.98% attention flops
jax.errors.SimplifiedTraceback: For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/maxtext/MaxText/train.py", line 776, in <module>
    app.run(main)
  File "/usr/local/lib/python3.10/dist-packages/absl/app.py", line 308, in run
    _run_main(main, args)
  File "/usr/local/lib/python3.10/dist-packages/absl/app.py", line 254, in _run_main
    sys.exit(main(argv))
  File "/opt/maxtext/MaxText/train.py", line 772, in main
    train_loop(config)
  File "/opt/maxtext/MaxText/train.py", line 666, in train_loop
    state, metrics = p_train_step(state, example_batch, nextrng)
jaxlib.xla_extension.XlaRuntimeError: INTERNAL: custom_partitioner: Traceback (most recent call last):
  File "/opt/maxtext/MaxText/train.py", line 776, in <module>
  File "/usr/local/lib/python3.10/dist-packages/absl/app.py", line 308, in run
  File "/usr/local/lib/python3.10/dist-packages/absl/app.py", line 254, in _run_main
  File "/opt/maxtext/MaxText/train.py", line 772, in main
  File "/opt/maxtext/MaxText/train.py", line 666, in train_loop
  File "/opt/jax/jax/_src/traceback_util.py", line 180, in reraise_with_filtered_traceback
  File "/opt/jax/jax/_src/pjit.py", line 338, in cache_miss
  File "/opt/jax/jax/_src/pjit.py", line 188, in _python_pjit_helper
  File "/opt/jax/jax/_src/core.py", line 2781, in bind
  File "/opt/jax/jax/_src/core.py", line 442, in bind_with_trace
  File "/opt/jax/jax/_src/core.py", line 948, in process_primitive
  File "/opt/jax/jax/_src/pjit.py", line 1737, in _pjit_call_impl
  File "/opt/jax/jax/_src/pjit.py", line 1713, in call_impl_cache_miss
  File "/opt/jax/jax/_src/pjit.py", line 1643, in _pjit_call_impl_python
  File "/opt/jax/jax/_src/interpreters/pxla.py", line 2321, in compile
  File "/opt/jax/jax/_src/interpreters/pxla.py", line 2835, in from_hlo
  File "/opt/jax/jax/_src/interpreters/pxla.py", line 2647, in _cached_compilation
  File "/opt/jax/jax/_src/compiler.py", line 427, in compile_or_get_cached
  File "/opt/jax/jax/_src/compiler.py", line 655, in _compile_and_write_cache
  File "/opt/jax/jax/_src/profiler.py", line 333, in wrapper
  File "/opt/jax/jax/_src/compiler.py", line 267, in backend_compile
  File "/opt/jax/jax/_src/custom_partitioning.py", line 155, in _custom_partitioning_partition
  File "/opt/transformer-engine/transformer_engine/jax/cpp_extensions/attention.py", line 1060, in partition
AssertionError: Sliding window attention is not supported when context parallelism is enabled
```

Comment (huanghua1994) [reply transformer_engine/jax/cpp_extensions/attention.py:1060]: It seems that the changes suggested by @zlsh80826 here https://github.com/NVIDIA/TransformerEngine/pull/1205#discussion_r1791106333 does not work. Probably this one should be the correct:
```python
is_context_parallel = get_mesh_axis_size(config.cp_axis, mesh) > 1
if is_context_parallel and config.window_size[0] > 0:
    assert (
        is_context_parallel and config.window_size[0] == -1
    ), "Sliding window attention is not supported when context parallelism is enabled"
```
So when `is_context_parallel  = False`, the assert will not be triggered.

Comment (huanghua1994) [reply transformer_engine/jax/cpp_extensions/attention.py:1060]: `not is_context_parallel or config.window_size[0] == -1` should be correct.

Comment (kocchop) [reply transformer_engine/jax/cpp_extensions/attention.py:1060]: so, the PR should be good, right? @huanghua1994

Comment (huanghua1994) [reply transformer_engine/jax/cpp_extensions/attention.py:1060]: Yes, I think the PR is good. Should be fine to merge after testing.

**Comment by xrennvidia** [comment]
I think this kind of bug could have been captured by the sliding window attention unit tests?  Do we have the tests in CI?

**Review by huanghua1994 — APPROVED** [review]

**Comment by huanghua1994** [comment]
> I think this kind of bug could have been captured by the sliding window attention unit tests? Do we have the tests in CI?

All CI tests passed before merging #1205. Probably there is no unit test that invokes CP and/or SWA.

**Comment by xrennvidia** [comment]
> > I think this kind of bug could have been captured by the sliding window attention unit tests? Do we have the tests in CI?
> 
> All CI tests passed before merging #1205. Probably there is no unit test that invokes CP and/or SWA.

If unit test does not invoke CP, it means `is_context_parallel=False`, which results in `assert False` and trigger the problem. Would be better if somebody can take a look at the test and make sure it can cover the bug. Thanks.

File: transformer_engine/jax/cpp_extensions/attention.py
Comment (zlsh80826) [reply transformer_engine/jax/cpp_extensions/attention.py:1060]: Oh yes, this is correct.

**Review by zlsh80826 — APPROVED** [review]

**Comment by zlsh80826** [comment]
> If unit test does not invoke CP, it means is_context_parallel=False, which results in assert False and trigger the problem. Would be better if somebody can take a look at the test and make sure it can cover the bug. Thanks.

This because the CP (and all distributed) testing is under L1 tests which are not testing for each PR. We probably need some sanity check for that as well on L0 tests. I have noted that and I will think about how to handle it well after I finish what I’m currently working on.

**Comment by xrennvidia** [comment]
> > If unit test does not invoke CP, it means is_context_parallel=False, which results in assert False and trigger the problem. Would be better if somebody can take a look at the test and make sure it can cover the bug. Thanks.
> 
> This because the CP (and all distributed) testing is under L1 tests which are not testing for each PR. We probably need some sanity check for that as well on L0 tests. I have noted that and I will think about how to handle it well after I finish what I’m currently working on.

Yeah, that would be better. Actually @kocchop encountered the issue while he run without CP, which I think should not be related to the CP tests that you are talking about.

**Review by ksivaman — APPROVED** [review]

**Comment by ksivaman** [comment]
@kocchop Thanks for the contribution! Could you please sign-off your commits using [this guide](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst#sign-your-work)?

**Comment by kocchop** [comment]
> @kocchop Thanks for the contribution! Could you please sign-off your commits using [this guide](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst#sign-your-work)?

@ksivaman done

**Comment by phu0ngng** [comment]
I tested this PR locally with the image `ghcr.io/nvidia/jax:pax-2024-10-02`. 
All tests in `test_distributed_fused_attn.py` passed, thus this PR can be merged.

**Pull request merged**
